### PR TITLE
feat: Added method_signature option to generated rpc methods

### PR DESCRIPTION
**Pull request by yon-mg**

### CODE REVIEW

File: src/main/java/com/google/cloud/discotoproto3converter/proto3/DocumentToProtoConverter.java [hunk 1]
```diff
@@ -314,15 +316,28 @@ private void readResources(Document document) {
...
           input.getFields().add(queryField);
           if (queryField.getValueType().isEnum()) {
             input.getEnums().add(queryField.getValueType());
           }
+
+          // optional method signatures
+          methodSignatureParams.add(queryField.getName());
         }
+
+        Option optionalMethodSignatureOption = new Option("google.api.method_signature");
```
Comment (vchudnov-g) [inline src/main/java/com/google/cloud/discotoproto3converter/proto3/DocumentToProtoConverter.java:338]: Let's have a plan for the optional signatures before we put in code to collect them. This new Option never seems to get read; let's not add dead code.

(You can insert a TODO, though, as a helpful pointer.)

File: src/main/java/com/google/cloud/discotoproto3converter/proto3/DocumentToProtoConverter.java [hunk 2]
```diff
@@ -351,6 +366,8 @@ private void readResources(Document document) {
         // Method
         GrpcMethod grpcMethod = new GrpcMethod(methodname, input, output, method.description());
         grpcMethod.getOptions().add(methodHttpOption);
+        grpcMethod.getOptions().add(requiredMethodSignatureOption);
```
Comment (vchudnov-g) [inline src/main/java/com/google/cloud/discotoproto3converter/proto3/DocumentToProtoConverter.java:369]: @vam-google Does discovery allow for required fields that are not path params? I would have expected so. Do you know? Are we assuming not? (I can modify my scripts to test existing Discovery files.)

File: src/main/java/com/google/cloud/discotoproto3converter/proto3/DocumentToProtoConverter.java [hunk 3]
```diff
@@ -351,6 +366,8 @@ private void readResources(Document document) {
...
+        // TODO: design heuristic for other useful mathod signatures with optional fields
```
Comment (vchudnov-g) [inline src/main/java/com/google/cloud/discotoproto3converter/proto3/DocumentToProtoConverter.java:370]: nit: sp: s/mathod/method/

File: src/main/java/com/google/cloud/discotoproto3converter/proto3/Proto3Writer.java [hunk 4]
```diff
@@ -107,13 +107,22 @@ private void printServices(Collection<GrpcService> services, PrintWriter writer)
...
-        StringBuilder optionsSb = new StringBuilder();
         for (Option option : method.getOptions()) {
-          optionsSb.append("    option (").append(option).append(") = {\n");
-          for (Map.Entry<String, String> prop : option.getProperties().entrySet()) {
-            optionsSb.append("      " + prop.getKey() + ": " + '"' + prop.getValue() + "\"\n");
-          }
+          StringBuilder optionsSb = new StringBuilder();
+          optionsSb.append("    option (").append(option).append(") = ");
+          
+          if (option.toString().equals("google.api.http")) {
```
Comment (vchudnov-g) [inline src/main/java/com/google/cloud/discotoproto3converter/proto3/Proto3Writer.java:114]: Now we're using strings for both `Option` construction in the other file and comparison here. I think it makes sense to use an enum of allowed `Option`s, and we refer to the enum value when creating and comparing the `Option` type. This allows the compiler to help us catch typos, and is more elegant. If you can do that now, great, if not, I suggest a TODO and we can do it in a subsequent PR.

**Review by vchudnov-g — COMMENTED** [review]
Looks good! Just a couple of comments. 

Also, do add a description to the PR. Say you're getting method signature from the required params, etc.

**Review by vam-google — COMMENTED** [review]
Overall looks good! 

Please use `"required: true"` parameter to determine what is required and try to reuse the option to string construction logic between methods and messages (check the specific comments for details).

**Comment by yon-mg** [comment]
@vchudnov-g @vam-google PTAL

**Review by vam-google — APPROVED** [review]
LGTM.

It would be still nice if you could put the option writer part into a private method and reuse it for message fields and rpc annotations. Please consider doing it before submitting this PR.

**Comment by yon-mg** [comment]
I did try to move that option writing code into a private method but it didn't seem to make things simpler or neater. Will discuss later since perhaps I was doing something roundabout.

File: src/main/java/com/google/cloud/discotoproto3converter/disco/Method.java [hunk 5]
```diff
@@ -64,6 +65,8 @@ public static Method from(DiscoveryNode root, Node parent) {
         pathParams.put(name, schema);
       } else if (schema.location().toLowerCase().equals("query")) {
         queryParams.put(name, schema);
+      } if (schema.required()) {
+        requiredParams.put(name, schema);
```
Comment (vchudnov-g) [inline src/main/java/com/google/cloud/discotoproto3converter/disco/Method.java:69]: nit: new line. This looks like an `else if`, even though it's not ;-)

File: src/main/java/com/google/cloud/discotoproto3converter/proto3/Proto3Writer.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (vchudnov-g) [reply src/main/java/com/google/cloud/discotoproto3converter/proto3/Proto3Writer.java:117]: were you able to do this? (sorry, lost track)

**Review by vchudnov-g — COMMENTED** [review]
looks fine. Just a minor point of order: it's good to wait for explicit Approval from all those you've assigned as reviewers and those who have already commented on the code, unless they say otherwise ;-)

**Pull request merged**